enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,303 @@
+version of the OpenShift variant I am communicating with so that I can
+use it to make decisions on the client side.
+
+As an application developer, I want to be able to determine the
+Kubernetes distribution the application is being installed into at
+runtime so that I only need to implement one set of installation
+logic.
+
+As an OpenShift user without admin privileges, I want to find
+information about the version of the product I am using in a
+consistent way, regardless of whether I am using standalone OpenShift,
+HyperShift, or MicroShift.
+
+### Goals
+
+* Use a standard Kubernetes type, to keep access easy for arbitrary
+  API clients. This also lets us avoid introducing a new CRD, so we do
+  not increase the overhead for MicroShift.
+* Define a standard location for version information that all
+  OpenShift variants can support.
+
+### Non-Goals
+
+* This change does not introduce a new way to update the version of a
+  cluster.
+* This change does not introduce a new way for a client to discover
+  that a cluster is being upgraded.
+* This change does not introduce a new way for a client to discover
+  the cluster topology (single-node, HA, external control-plane, etc.).
+
+## Proposal
+
+Because we want arbitrary clients to be able to read and understand
+the data, we will use a standard Kubernetes ConfigMap.
+
+Because we want all OpenShift variants to publish the information in
+the same way and that the data should not be writable by API callers,
+we will specify the name and namespace for the ConfigMap.
+
+Because MicroShift exposes a limited number of namespaces by default,
+we need to choose from one of those. The most suitable available
+namespace is `openshift`, so we will create the ConfigMap with name
+`version` in the `openshift` namespace.
+
+This choice will mean updating the implementation in MicroShift to
+change the namespace from `kube-public`, but since the ConfigMap will
+be OpenShift-specific we want to avoid collisions that might be
+introduced by other distributions in the `kube-public` namespace.
+
+Because we want all OpenShift variants to report the same type of
+information, we will keep the contents of the ConfigMap brief and
+specify the data key names in this enhancement:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: version
+  namespace: openshift
+data:
+  product: ""
+  major: ""
+  minor: ""
+  version: ""
+```
+
+- `product` should hold a representation of the OpenShift variant
+  name. For example, `OCP` for OpenShift in standalone or external
+  control-plane configurations of all topologies and `MicroShift` for
+  MicroShift instances.
+- `major` should hold the major version number of the product as a
+  string. For example, for `4.12.0` the value would be `"4"`.
+- `minor` should hold the minor version number of the product as a
+  string. For example, for `4.12.0` the value would be `"12"`.
+- `version` should hold the complete version of the product as a
```

Comment:
Do all products that might be using this ConfigMap use [SemVer][1]?  Because if so, seems like we should commit to that here.  And also like we could avoid splitting out `major` and `minor`, and leave that as an excercise for interested consumers to parse out of the SemVer `version`.  Slightly more work for those consumers, but also no risk of having internally-inconsistent data in the in-cluster ConfigMap.

[1]: https://semver.org/spec/v2.0.0.html

Reply:
I don't see any real challenge with ensuring the information is consistent. Behavior shouldn't change based on a patch level, but if we find that's not the case we can add a patch version field in the future.

Reply:
Yeah, I am not calling for "conveniently expose `patch`".  I'm saying that conveniently exposing `major` and `minor` is a small convenience that brings along a small risk.  I'm fine if folks feel like the convenience still outweighs the risk.

Reply:
Unless Microshift isn't willing to constrain version to those parseable by semver I'd prefer we just rely on semver parsing.

Reply:
It was mentioned on Slack that Microshift will always have the same versions as OCP. Which is slightly surprising because that constrains them to only being able to deliver microshift fixes when we deliver a new version of OCP, but assuming this holds true it seems like we're assured that only semver versions will be used.

Reply:
As mentioned in the [original MicroShift enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/microshift/kubernetes-for-device-edge.md):

> MicroShift is built downstream of OpenShift to ensure that it includes the same versions of the embedded components that are present in standalone OpenShift, allowing us to maintain binary compatibility.